torch/csrc/jit/codegen/cuda/glfdc/eval.h

```diff
@@ -0,0 +1,283 @@
+  }
+};
+
+// SymbolicExpr - symbolic expression
+//
+// This is stack machine that is holding expression in form of expression
+// reduction tree. All operations to calculate expression value are kept in
+// linear tree form to allow lazy evaluation.
+//
+// On creation of expression evaluator initial stack of known (scalar values is
+// prepared and memoized) During evaluation runtime (unknown) values are being
+// queried and inserted to the stack copy Afterwards expression is calculated by
+// DFS right first tree reduction using stack algorithm
+//
+// Symbolic expression can be later on translated to "optimized" form for faster
+// evaluation. This is possible once all expressions belonging to ExprDag are
+// already created. This translation requires knowledge of all subexpressions
+// reuses, thus no additional subexpression can be added (result of such
+// calculation would not be memoized).
+class TORCH_CUDA_CU_API SymbolicExpr {
+  // stores where to update on stack runtime value
+  using initial_stack_t = InitialStack<scalar_type>;
+  using stack_t = EvalStack<scalar_type>;
+
+ public:
+  // User provided function that returns concrete value for symbolic value
+  // Argument to this function is same unique value for which the binding was
+  // created. See ExpressionBuilder::getBinding();
+  using binding_fn_t = const BoundValueProvider&;
+
+  // Template of symbolic expression
+  struct Template {
+    // List of operations to perform to evaluate an expression stored in
+    // linearized tree form
+    std::vector<Operation> operations;
+    // initial stack of known operands with gaps for runtime values
+    initial_stack_t initial_stack;
+    // List of gap indices on initial_stack for runtime values
+    // Those are filled in on copy of the stack during evaluation
+    std::vector<BindingGapType> binding_gaps;
+  };
+
+  explicit SymbolicExpr(const Expr& e);
+
+  // Evaluates a symbolic expression.
+  template <typename Ty_>
+  opt_scalar_t evaluate(
+      EvalState& es,
+      const SymbolicValueProviderBase<Ty_>& symbols) const {
+    return evaluateExpr(es, symbols.getProvider());
+  }
+
+  // Returns reference to owner DAG
+  const ExprDAG& dag() const;
+
+  // Returns reference to expression in ExprDAG
+  const Expr& expr() const;
+
+  // Translates expression to optimized form.
+  //
+  // NOTE: This should be done only after all expressions has been added to
+  // ExprDAG
+  void optimize();
+
+  // Returns true if expression wasn't translated to optimized form
+  bool isOptimized() const noexcept {
+    return program_.isValid();
+  }
+
+ private:
+  opt_scalar_t evaluateExpr(EvalState& e, binding_fn_t binding_fn) const;
+
+  void evaluateSubExpr(size_t op_index, stack_t& stack, EvalState& es) const;
+
+  // Evaluates expression in optimized form
+  opt_scalar_t evaluateOptimized(EvalState& es, binding_fn_t binding_fn) const;
+  // Evaluates expression from template (fallback)
```

Comment:
When is template used? Looks like if the optimized form doesn't have the value we want. Would we be able to go back to legacy pointer chasing method on the IR nodes if that's the case?